### PR TITLE
Allow missing trait members assist without needing braces

### PR DESCRIPTION
**Pull request by M-J-Hooper**
Assist to complete missing items when implementing a trait does not appear without impl def braces (see #5144 ).

The reason behind this was that this assist is based on `ast::AssocItemList` which only appears in the AST after the braces are added to the impl def.

Instead of relying on and replacing the item list, we now instead replace the entire `ast::Impl` and add the item list if its missing.

### CODE REVIEW

File: crates/assists/src/handlers/add_missing_impl_members.rs [hunk 1]
```diff
@@ -148,11 +146,14 @@ fn add_missing_impl_members_inner(
...
+        let impl_item_list = impl_def.assoc_item_list().unwrap_or(make::assoc_item_list());
+
         let n_existing_items = impl_item_list.assoc_items().count();
         let source_scope = ctx.sema.scope_for_def(trait_);
-        let target_scope = ctx.sema.scope(impl_item_list.syntax());
+        let target_scope = ctx.sema.scope(impl_def.syntax());
         let ast_transform = QualifyPaths::new(&target_scope, &source_scope)
-            .or(SubstituteTypeParams::for_trait_impl(&source_scope, trait_, impl_def));
+            .or(SubstituteTypeParams::for_trait_impl(&source_scope, trait_, impl_def);
+
```
Comment (lnicola) [inline crates/assists/src/handlers/add_missing_impl_members.rs:156]: Missing )

**Review by lnicola — APPROVED** [review]
Looks good to me, thanks!

File: crates/assists/src/handlers/add_missing_impl_members.rs (hunk 1 above)
Comment (M-J-Hooper) [reply crates/assists/src/handlers/add_missing_impl_members.rs:156]: Aaand that's why you shouldn't push at 1am... Fixed.

File: crates/syntax/src/ast/edit.rs [hunk 2]
```diff
@@ -93,6 +93,22 @@ where
     }
 }
 
+impl ast::Impl {
+    #[must_use]
+    pub fn with_items(&self, items: ast::AssocItemList) -> ast::Impl {
```
Comment (matklad) [inline crates/syntax/src/ast/edit.rs:98]: Let's be maximally boring here: `with_assoc_item_list`. One day, we'll auto-generate this....

**Comment by matklad** [comment]
r=me with method renamed!

bors d+

**Comment by matklad** [comment]
bors r+

Thanks! (you don't need to waint until CI is green before approving, the robot can do that)

**Pull request merged**